tensorflow/compiler/mlir/xla/transforms/tensor_linalg_to_buffer_linalg.cc

```diff
@@ -0,0 +1,143 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for transforming Linalg operations with tensor
+// types to memref type and allocate and deallocate buffers using the
+// BufferAssignmentPlacer.
+
+#include "absl/memory/memory.h"
+#include "llvm/ADT/APInt.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"    // TF:llvm-project
+#include "mlir/Dialect/Linalg/IR/LinalgTypes.h"  // TF:llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"     // TF:llvm-project
+#include "mlir/IR/AffineExpr.h"                  // TF:llvm-project
+#include "mlir/IR/Attributes.h"                  // TF:llvm-project
+#include "mlir/IR/Builders.h"                    // TF:llvm-project
+#include "mlir/IR/Function.h"                    // TF:llvm-project
+#include "mlir/IR/Location.h"                    // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"                 // TF:llvm-project
+#include "mlir/IR/Operation.h"                   // TF:llvm-project
+#include "mlir/IR/PatternMatch.h"                // TF:llvm-project
+#include "mlir/IR/StandardTypes.h"               // TF:llvm-project
+#include "mlir/Pass/Pass.h"                      // TF:llvm-project
+#include "mlir/Transforms/DialectConversion.h"   // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/buffer_assignment.h"
+#include "tensorflow/compiler/mlir/xla/transforms/rewriters.h"
+
+namespace mlir {
+namespace xla {
+namespace {
+
+class GenericOpConverter
+    : public xla::BufferAssignmentOpConversionPattern<linalg::GenericOp> {
+ public:
+  using xla::BufferAssignmentOpConversionPattern<
+      linalg::GenericOp>::BufferAssignmentOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp op, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const final {
+    auto loc = op.getLoc();
+    SmallVector<Value, 4> args(operands.begin(), operands.end());
+
+    // Update all types to memref types.
+    auto result = op.getOperation()->getResult(0);
+    auto type = result.getType().cast<ShapedType>();
+    auto memref_type = MemRefType::get(type.getShape(), type.getElementType());
+    auto position = bufferAssignment->computeAllocPosition(result);
+
+    // Compute alloc position and insert a custom allocation node.
+    OpBuilder allocBuilder(result.getDefiningOp());
+    allocBuilder.restoreInsertionPoint(position);
+    auto alloc = allocBuilder.create<AllocOp>(loc, memref_type);
+    args.push_back(alloc);
+
+    // Generate a new linalg operation that works on buffers.
+    auto linalgOp = rewriter.create<linalg::GenericOp>(
+        loc, llvm::None, args, rewriter.getI64IntegerAttr(operands.size()),
+        rewriter.getI64IntegerAttr(1), op.indexing_maps(), op.iterator_types(),
+        /*doc=*/nullptr,
```

Comment:
can you carry the extra parameters over here? Or at least assert that they are `nullptr` before, as well.

Reply:
Carried over.